lib/AST/DocComment.cpp

```diff
@@ -426,24 +426,44 @@ class CommentProviderFinder final {
     return std::nullopt;
   }
 
+  /// Check if there is an inherited protocol that has a default implementation
+  /// of `VD` with a doc comment.
   std::optional<ResultWithDecl> findDefaultProvidedDecl(const ValueDecl *VD) {
-    // Only applies to protocol extension member.
-    auto *protocol = VD->getDeclContext()->getExtendedProtocolDecl();
-    if (!protocol)
+    NominalTypeDecl *nominalType =
+        dyn_cast_or_null<NominalTypeDecl>(VD->getDeclContext()->getAsDecl());
+    if (!nominalType) {
+      nominalType = VD->getDeclContext()->getExtendedProtocolDecl();
+    }
+    if (!nominalType)
       return std::nullopt;
 
     SmallVector<ValueDecl *, 2> members;
-    protocol->lookupQualified(const_cast<ProtocolDecl *>(protocol),
-                              DeclNameRef(VD->getName()),
-                              VD->getLoc(), NLOptions::NL_ProtocolMembers,
-                              members);
+    nominalType->lookupQualified(nominalType, DeclNameRef(VD->getName()),
+                                 VD->getLoc(), NLOptions::NL_ProtocolMembers,
+                                 members);
```

Comment:
Thought of a couple more edge cases :)

1. static mismatch

```swift
protocol P {}
extension P {
  /// hello
  static func foo() {}
}
struct S: P {
  func foo() {}
}
func bar(_ x: S) {
  x.foo()
}
```

2. enum case matching a function (if we want this to work it would need to be a static function)

```swift
protocol P {}
extension P {
  /// hello
  func foo(_ x: Int) -> E { .foo(0) }
}

enum E: P {
  case foo(Int)
}

func bar(_ x: E) {
  E.foo(0)
}
```

3. this is quite silly, but technically this would match:

```swift
class C {
  /// hello
  struct foo {}
}
class D: C {
  var foo: C.foo.Type { C.foo.self }
}

func bar(_ x: D) {
  let y = x.foo
}
```

Reply:
Good cases, thanks. Case 3 does inherit but I decided to not care and not add a test case for it. If it broke, I don’t think we would care.